src/readers/src/main/scala/BinaryFileFormat.scala

```diff
@@ -0,0 +1,215 @@
+import org.apache.commons.io.{FilenameUtils, IOUtils}
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs.{FileStatus, Path}
+import org.apache.hadoop.io.BytesWritable
+import org.apache.hadoop.mapreduce._
+import org.apache.hadoop.mapreduce.lib.input.{FileInputFormat, FileSplit}
+import org.apache.hadoop.mapreduce.task.TaskAttemptContextImpl
+import org.apache.log4j.Logger
+import org.apache.spark.TaskContext
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.GenericInternalRow
+import org.apache.spark.sql.execution.datasources._
+import org.apache.spark.sql.sources._
+import org.apache.spark.sql.types._
+import org.apache.spark.unsafe.types.UTF8String
+import org.apache.spark.util.SerializableConfiguration
+
+import scala.util.Random
+
+/**
+  * Actually reads the records from files
+  *
+  * @param subsample  what ratio to subsample
+  * @param inspectZip whether to inspect zip files
+  */
+private[spark] class BinaryRecordReader(val subsample: Double, val inspectZip: Boolean)
+  extends RecordReader[String, BytesWritable] {
+
+  private var done: Boolean = false
+  private var inputStream: InputStream = _
+  private var filename: String = _
+  private var recordValue: BytesWritable = _
+  private var progress: Float = 0.0F
+  private val rng: Random = new Random()
+  private var zipIterator: ZipIterator = _
+
+  override def close(): Unit = {
+    if (inputStream != null) {
+      inputStream.close()
+    }
+  }
+
+  override def getCurrentKey: String = {
+    filename
+  }
+
+  override def getCurrentValue: BytesWritable = {
+    recordValue
+  }
+
+  override def getProgress: Float = {
+    progress
+  }
+
+  override def initialize(inputSplit: InputSplit, context: TaskAttemptContext): Unit = {
+    // the file input
+    val fileSplit = inputSplit.asInstanceOf[FileSplit]
+
+    // the actual file we will be reading from
+    val file = fileSplit.getPath
+    // job configuration
+    val conf = context.getConfiguration
+    // get the filesystem
+    val fs = file.getFileSystem(conf)
+    // open the File
+    filename = file.toString
+
+    inputStream = fs.open(file)
+    rng.setSeed(filename.hashCode.toLong)
+    if (inspectZip && FilenameUtils.getExtension(filename) == "zip") {
+        zipIterator = new ZipIterator(inputStream, filename, subsample, rng)
+    }
+  }
+
+  def markAsDone(): Unit = {
+    done = true
+    progress = 1.0F
+  }
+
+  override def nextKeyValue(): Boolean = {
+    if (done) {
+      return false
+    }
+
+    if (zipIterator != null) {
+      if (zipIterator.hasNext) {
+        val (fn, barr) = zipIterator.next
+        filename = fn
+        recordValue = new BytesWritable()
+        recordValue.set(barr, 0, barr.length)
+        true
+      } else {
+        markAsDone()
+        false
+      }
+    } else {
+      rng.setSeed(filename.hashCode.toLong)
```

Comment:
We should not be mutating the RNG that's passed in (fine, I agree generating a number mutates a state but setting the seed inside an iterator in a way that's not controllable from outside seems problematic). The seed should be set outside.

Reply:
Also, please open an item to factor into a subclass for zip files. I'm not really sure how exactly do structure the code but that seems slightly more robust than if-elses.

Reply:
A more real problem with this is that it sets the seed to the hash, which kills the randomality of any following random calls, no?